src/Type/Php/ArrayAccessOffsetExistsMethodTypeSpecifyingExtension.php

```diff
@@ -0,0 +1,96 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Php;
+
+use ArrayAccess;
+use PhpParser\Node\Expr\MethodCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Analyser\SpecifiedTypes;
+use PHPStan\Analyser\TypeSpecifier;
+use PHPStan\Analyser\TypeSpecifierAwareExtension;
+use PHPStan\Analyser\TypeSpecifierContext;
+use PHPStan\Reflection\MethodReflection;
+use PHPStan\Reflection\ParametersAcceptorSelector;
+use PHPStan\Reflection\ReflectionProvider;
+use PHPStan\Type\Accessory\HasOffsetType;
+use PHPStan\Type\Accessory\HasOffsetValueType;
+use PHPStan\Type\Constant\ConstantIntegerType;
+use PHPStan\Type\Constant\ConstantStringType;
+use PHPStan\Type\Generic\GenericObjectType;
+use PHPStan\Type\MethodTypeSpecifyingExtension;
+use PHPStan\Type\TypeCombinator;
+use function count;
+
+final class ArrayAccessOffsetExistsMethodTypeSpecifyingExtension implements MethodTypeSpecifyingExtension, TypeSpecifierAwareExtension
+{
+
+	private TypeSpecifier $typeSpecifier;
+
+	public function __construct(
+		private ReflectionProvider $reflectionProvider,
+	)
+	{
+	}
+
+	public function setTypeSpecifier(TypeSpecifier $typeSpecifier): void
+	{
+		$this->typeSpecifier = $typeSpecifier;
+	}
+
+	public function getClass(): string
+	{
+		return ArrayAccess::class;
+	}
+
+	public function isMethodSupported(
+		MethodReflection $methodReflection,
+		MethodCall $node,
+		TypeSpecifierContext $context,
+	): bool
+	{
+		return $methodReflection->getName() === 'offsetExists' && $context->true();
+	}
+
+	public function specifyTypes(MethodReflection $methodReflection, MethodCall $node, Scope $scope, TypeSpecifierContext $context): SpecifiedTypes
+	{
+		if (count($node->getArgs()) < 1) {
+			return new SpecifiedTypes();
+		}
+		$key = $node->getArgs()[0]->value;
+		$keyType = $scope->getType($key);
+
+		if (
+			!$keyType instanceof ConstantStringType
+			&& !$keyType instanceof ConstantIntegerType
+		) {
+			return new SpecifiedTypes();
+		}
+
+		foreach($scope->getType($node->var)->getObjectClassReflections() as $classReflection) {
+			$implementsTags = $classReflection->getImplementsTags();
+
+			if (
+				!isset($implementsTags[\ArrayAccess::class])
+				|| !$implementsTags[\ArrayAccess::class]->getType() instanceof GenericObjectType
+			) {
+				continue;
+			}
+
+			$implementsType = $implementsTags[\ArrayAccess::class]->getType();
+			$arrayAccessGenericTypes = $implementsType->getTypes();
+			if (!isset($arrayAccessGenericTypes[1])) {
+				continue;
+			}
+
+			return $this->typeSpecifier->create(
+				$node->var,
+				new HasOffsetValueType($keyType, $arrayAccessGenericTypes[1]),
+				$context,
+				$scope,
+			);
+		}
```

Comment:
reading the "raw" array-access generic type, because `ObjectType->getValueType` will append `null` for possibly non-existant offsets. the added `null` makes it impossible for us to differentiate a nullable-typed generic from a non-nullable generic